src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -692,6 +692,25 @@ bool
 fsm_stop_postgres(Keeper *keeper)
 {
 	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	PGSQL *pgsql = &(postgres->sqlClient);
+
+	if (pg_setup_is_running(pgSetup))
+	{
+		/*
+		 * The first checkpoint writes all the in-memory buffers, the second
+		 * checkpoint writes everything that was added during the first one. The
+		 * third one might have still some work to do.
+		 */
+		for (int i = 0; i < 3; i++)
+		{
+			if (!pgsql_checkpoint(pgsql))
```

Comment:
When Postgres stops it does a checkpoint anyway. See  comments in https://github.com/postgres/postgres/blob/master/src/backend/postmaster/checkpointer.c#L13 and here:

```c
 * The checkpointer is started by the postmaster as soon as the startup
 * subprocess finishes, or as soon as recovery begins if we are doing archive
 * recovery.  It remains alive until the postmaster commands it to terminate.
 * Normal termination is by SIGUSR2, which instructs the checkpointer to
 * execute a shutdown checkpoint and then exit(0).  (All backends must be
 * stopped before SIGUSR2 is issued!)  Emergency termination is by SIGQUIT;
 * like any backend, the checkpointer will simply abort and exit on SIGQUIT.
```

The advantage of doing a CHECKPOINT ourselves is that we do that while online, and then the shutdown sequence of Postgres can be fast. I'm not completely sold on the loop-of-3 thing, but certainly reducing the time when Postgres has not shut down yet but can't do anything anymore is a good thing.

Reply:
Changed it to doing 2 checkpoints before stopping Postgres. I also made it so that if we fail the second checkpoint, we redo only that one rather than two of them again.

Reply:
> but certainly reducing the time when Postgres has not shut down yet but can't do anything anymore is a good thing.

I thought so to, but I don't think anymore this is the case for all our shutdowns. For this maintenance case it definitely seems good. However, when the primary notices it's on the losing side of a network partition it must stop accepting writes quickly. Otherwise there can be a split brain. So waiting for a CHECKPOINT to finish (even a single one), before stopping to accept writes seems bad in that case.

I think this makes it complex enough to warrant a separate PR.

Reply:
Resolving this since it's now only called when going to maintenance.